File: 502-ipo/ipo.py
class Solution:
    def findMaximizedCapital(self, k, w, pr, cp):
        n = len(pr)
        projects = sorted(zip(cp, pr), key=lambda x: x[0])
        
        min_heap = []
        i = 0
        while k > 0:
            while i < n and w >= projects[i][0]:
                heapq.heappush(min_heap, -projects[i][1])
                i += 1
            if min_heap:
                w += -heapq.heappop(min_heap)
            k -= 1
        
        return w

File: 502-ipo/ipo.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] pr, int[] cp) {
        int n = pr.length;
        List<Project> projects = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            projects.add(new Project(cp[i], pr[i]));
        }
        Collections.sort(projects, (a, b) -> a.cost - b.cost);
        
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        int i = 0;
        while (k > 0) {
            while (i < n && w >= projects.get(i).cost) {
                minHeap.offer(-projects.get(i).profit);
                i++;
            }
            if (!minHeap.isEmpty()) {
                w += -minHeap.poll();
            }
            k--;
        }
        
        return w;
    }
    
    private static class Project {
        int cost;
        int profit;
        
        Project(int cost, int profit) {
            this.cost = cost;
            this.profit = profit;
        }
    }
}

